### PR TITLE
Improve delay in accessing WebDAV folders on external storage

### PR DESCRIPTION
**Pull request by luka-nextcloud**
* Resolves: # 

## Summary
**Issue**: loading time while open an external directory increases with the number of children (files, folders).
**Root cause**: profind function has been called for each child to fetch metadata/permissions and it seem unneccessary.
**Update**: 
- Caching the profind response of children while calling function opendir. Then, we don't need to call profind function for each child anymore.
- Optimize propfind caching to prevent duplicate call

**Before**: calling of profind function for each child increases loading time; propfind duplicate call for parent dir
image

**After**: calling profind function reduced since profind response has been cached while calling opendir; no duplicate propfind call
image

## TODO

- [ ] ...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: lib/private/Files/Storage/DAV.php [hunk 1]
```diff
@@ -254,7 +263,9 @@ public function opendir($path) {
 				if (!$this->statCache->hasKey($path)) {
 					$this->statCache->set($file, true);
 				}
+				$fileDetail = $response[$file];
 				$file = basename($file);
+				$this->statCache->set($file, $fileDetail);
```
[the remark this answers is not in the document]
Comment (luka-nextcloud) [reply lib/private/Files/Storage/DAV.php:268]: @juliushaertl I don't think we really need a separate cache since function profind also uses stateCache to cache the profind response.

File: core/css/publicshareauth.css [hunk 2]
```diff
@@ -1,36 +1,36 @@
...
-	width: 45px !important;
-	height: 45px;
-	background-color: transparent !important;
-}
-
-.warning > .warning {
-	/* Do not use a top margin for warning messages in the warning container. */
-	margin-top: 0;
-}
+form fieldset {
```
Comment (juliushaertl) [inline core/css/publicshareauth.css:1]: Seems unrelated, let's get rid of this change 😁

**Comment by juliushaertl** [comment]
Since you're at it, maybe you can also look into saving the remaining duplicate request that is still left:

Screenshot 2023-05-29 at 11 28 12

**Comment by juliushaertl** [comment]
Can you double check that? When I was testing it the propfind was still duplicate as the second one didn't read the cached response from anywhere. 

I think it would make sense to reuse the internal `profind` method in https://github.com/nextcloud/server/pull/38418/files#diff-53ee4c908f9ae3a27e45d03c57ac5ed83c5916d7260561a03f279880671f6155R240 instead of directly calling the `$this->client->propFind`

**Comment by luka-nextcloud** [comment]
> Can you double check that? When I was testing it the propfind was still duplicate as the second one didn't read the cached response from anywhere.

You are right, I updated the screenshot again.

> I think it would make sense to reuse the internal profind method in https://github.com/nextcloud/server/pull/38418/files#diff-53ee4c908f9ae3a27e45d03c57ac5ed83c5916d7260561a03f279880671f6155R240 instead of directly calling the $this->client->propFind

As I know, we have to use $this->client->propFind() since the target item (file, folder) is an external dir. The internal propFind is only for internal item (file,folder)

**Review by juliusknorr — APPROVED** [review]
Tested and works

**Comment by juliushaertl** [comment]
I've pushed a small fixup commit to address an encoding issue with files with spaces, but otherwise this seemed good.

Note that I'll revoke my initial question for reusing the first propfind as i think this one is intentional only done on the root without children (depth=0) as it would also get requested when listing the parent folder of the external storage mount. In this case it might be wise to only request relevant information.

<details>
<summary>In case we want to optimize that this could easily be done through the following patch</summary>

```diff
diff --git a/lib/private/Files/Storage/DAV.php b/lib/private/Files/Storage/DAV.php
index 267f32a0e5b..cf031ac8c00 100644
--- a/lib/private/Files/Storage/DAV.php
+++ b/lib/private/Files/Storage/DAV.php
@@ -314,9 +314,11 @@ class DAV extends Common {
                                                '{DAV:}getetag',
                                                '{DAV:}quota-available-bytes',
                                        ]
-                               );
+                               , 1);
                                $this->statCache->set($path, $response);
-                               $this->propfindCache->set($path, $response);
+                               foreach ($response as $file => $responseData) {
+                                       $this->propfindCache->set($file, $responseData);
+                               }
                        } catch (ClientHttpException $e) {
                                if ($e->getHttpStatus() === 404 || $e->getHttpStatus() === 405) {
                                        $this->statCache->clear($path . '/');
```

**Comment by icewind1991** [comment]
I'm a bit hesitant (but not fully against) adding extra caching layers.

The issue shown in the OP should also be solvable by implementing a more optimized version of `IStorage::getDirectoryContent` for `Dav`.

The default implementation that is currently inherited is doing the "stat every file in the folder" that seems to be causing the issue. Changing it to "load the folder once and the metadata for each item" should give the same speedup in that case.

**Comment by come-nc** [comment]
> I'm a bit hesitant (but not fully against) adding extra caching layers.
> 
> The issue shown in the OP should also be solvable by implementing a more optimized version of `IStorage::getDirectoryContent` for `Dav`.
> 
> The default implementation that is currently inherited is doing the "stat every file in the folder" that seems to be causing the issue. Changing it to "load the folder once and the metadata for each item" should give the same speedup in that case.

But the caching may help with other operations as well no?

**Comment by juliushaertl** [comment]
I think the caching here makes sense for all cases here also if it is not just about iterating over the directory content. It especially helps as the previous code called a propfind on any call that would obtain file metadata like the getPermission which is likely to be called in a few other places, so I think caching in an ArrayCache for the scope of the current request is still the most sane approach here.

**Review by come-nc — APPROVED** [review]

**Comment by Alexandero89** [comment]
Don´t know if its because of your commits, but i was testing your PR and checked the logs.
I get some strange absolute urls there.
Example log:
`{"reqId":"jXHhrkauwfkRmocMZ0C5","level":0,"time":"2023-06-19T11:03:05+00:00","remoteAddr":"","user":"--","app":"dav","method":"","url":"--","message":"sending dav PROPFIND request to external storage: http://localhosthttps://www.mycloud.com/remote.php/webdav/Backups/Android/Download/DCIM/OpenCamera/IMG_20210313_100632.jpg","userAgent":"--","version":"26.0.2.1","data":{"app":"dav"}}`

Should be: "https://www.mycloud.com/remote.php/webdav/Backups/Android/Download/DCIM/OpenCamera/IMG_20210313_100632.jpg"

Is: "http://localhosthttps://www.mycloud.com/remote.php/webdav/Backups/Android/Download/DCIM/OpenCamera/IMG_20210313_100632.jpg"

So "$request->getAbsoluteUrl()" is returning something strange.

**Comment by juliushaertl** [comment]
Let's close this then as https://github.com/nextcloud/server/pull/38945 was merged
